Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp

```diff
@@ -0,0 +1,302 @@
+                                                              std::move(color_key));
+    }
+
+    // Default to the default controls but overwrite if the creator
+    // has provided something specific
+    picojson::object host_controls = default_host_controls;
+    const picojson::value& host_controls_json = data.get("host_controls");
+    if (host_controls_json.is<picojson::object>())
+    {
+      host_controls = host_controls_json.get<picojson::object>();
+    }
+
+    if (host_controls.empty())
+    {
+      ERROR_LOG(VIDEO,
+                "Failed to load dynamic input json file '%s' because field "
+                "'host_controls' is missing ",
+                json_file.c_str());
+      m_valid = false;
+      return;
+    }
+
+    for (auto& [host_control, image_name] : host_controls)
+    {
+      texture_data.m_host_input_to_input_image.insert_or_assign(host_control, image_name.to_str());
+    }
+
+    // Default to the default device but overwrite if the creator
+    // has provided something specific
+    texture_data.m_devicename = default_device;
+    const picojson::value& device_json = data.get("device");
+    if (device_json.is<std::string>())
+    {
+      texture_data.m_devicename = device_json.get<std::string>();
+    }
+
+    if (texture_data.m_devicename.empty())
+    {
+      ERROR_LOG(VIDEO,
+                "Failed to load dynamic input json file '%s' because field "
+                "'device' is missing ",
+                json_file.c_str());
+      m_valid = false;
+      return;
+    }
+
+    m_dynamic_input_textures.emplace_back(std::move(texture_data));
+  }
+}
+
+DynamicInputTextureConfiguration::~DynamicInputTextureConfiguration() = default;
+
+void DynamicInputTextureConfiguration::GenerateTextures(const IniFile::Section* sec) const
+{
+  bool any_dirty = false;
+  for (const auto& texture_data : m_dynamic_input_textures)
+  {
+    any_dirty |= GenerateTexture(sec, texture_data);
+  }
+
+  if (!any_dirty)
+  {
+    return;
+  }
+  if (Core::GetState() == Core::State::Starting)
+  {
+    return;
+  }
+  if (!g_renderer)
+  {
+    return;
+  }
+  g_renderer->ForceReloadTextures();
+}
+
+bool DynamicInputTextureConfiguration::GenerateTexture(
+    const IniFile::Section* sec, const DynamicInputTextureData& texture_data) const
+{
+  std::string device_name;
+  if (!sec->Get("Device", &device_name) || device_name != texture_data.m_devicename)
+  {
+    return false;
+  }
+
+  // Load colorkey image copy
+  auto color_key_image = LoadImage(m_base_path + texture_data.m_colorKeyImageName);
+  bool dirty = false;
+  for (auto& [emulated_key, color] : texture_data.m_emulated_input_to_color)
+  {
+    std::string host_key = "";
+    sec->Get(emulated_key, &host_key);
+
+    const auto color_only_mask = MaskColor(*color_key_image, Pixel{color.r, color.g, color.b, 255});
+    const auto rects = FindRectRegionsOfNeighboringPixels(color_only_mask);
```

Comment:
You apply the copy operation to all found rectangles regardless of size.
I guess uses must ensure not one pixel of their chosen color is found in other parts of the image?

Reply:
Multiple replacements are a convenience at times.  I've seen a number of games that use the same sub-image (say an 'A' button icon) in multiple places in a single texture.